{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ignore wanring\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Basic process\n",
    "import pandas as pd\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Load Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tissue Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_title</th>\n",
       "      <th>dataset</th>\n",
       "      <th>status</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age</th>\n",
       "      <th>stage</th>\n",
       "      <th>tnm.t</th>\n",
       "      <th>tnm.n</th>\n",
       "      <th>tnm.m</th>\n",
       "      <th>tumor.location</th>\n",
       "      <th>...</th>\n",
       "      <th>tp53.mutation.protein</th>\n",
       "      <th>kras.mutation</th>\n",
       "      <th>kras.mutation.dna</th>\n",
       "      <th>kras.mutation.exon.number</th>\n",
       "      <th>kras.mutation.protein</th>\n",
       "      <th>braf.mutation</th>\n",
       "      <th>braf.mutation.dna</th>\n",
       "      <th>braf.mutation.exon.number</th>\n",
       "      <th>braf.mutation.protein</th>\n",
       "      <th>cit.molecularsubtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM971957</th>\n",
       "      <td>CIT001</td>\n",
       "      <td>discovery</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>Male</td>\n",
       "      <td>34.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T4</td>\n",
       "      <td>N2</td>\n",
       "      <td>M1</td>\n",
       "      <td>distal</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>M</td>\n",
       "      <td>c.35G&gt;A</td>\n",
       "      <td>c.35G&gt;A</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971958</th>\n",
       "      <td>CIT002</td>\n",
       "      <td>discovery</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>Female</td>\n",
       "      <td>36.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T3</td>\n",
       "      <td>N2</td>\n",
       "      <td>M1</td>\n",
       "      <td>distal</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971959</th>\n",
       "      <td>CIT003</td>\n",
       "      <td>discovery</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>Male</td>\n",
       "      <td>36.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>T4</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>proximal</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971960</th>\n",
       "      <td>CIT004</td>\n",
       "      <td>discovery</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>Female</td>\n",
       "      <td>38.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>T2</td>\n",
       "      <td>N0</td>\n",
       "      <td>M0</td>\n",
       "      <td>distal</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971961</th>\n",
       "      <td>CIT005</td>\n",
       "      <td>discovery</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>Female</td>\n",
       "      <td>39.2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T3</td>\n",
       "      <td>N1</td>\n",
       "      <td>M1</td>\n",
       "      <td>distal</td>\n",
       "      <td>...</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>N/A</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      !Sample_title    dataset              status     Sex  \\\n",
       "!Sample_geo_accession                                                        \n",
       "GSM971957                    CIT001  discovery   colorectal cancer    Male   \n",
       "GSM971958                    CIT002  discovery   colorectal cancer  Female   \n",
       "GSM971959                    CIT003  discovery   colorectal cancer    Male   \n",
       "GSM971960                    CIT004  discovery   colorectal cancer  Female   \n",
       "GSM971961                    CIT005  discovery   colorectal cancer  Female   \n",
       "\n",
       "                        age  stage tnm.t tnm.n tnm.m tumor.location  ...  \\\n",
       "!Sample_geo_accession                                                ...   \n",
       "GSM971957              34.5    4.0    T4    N2    M1         distal  ...   \n",
       "GSM971958              36.4    4.0    T3    N2    M1         distal  ...   \n",
       "GSM971959              36.9    2.0    T4    N0    M0       proximal  ...   \n",
       "GSM971960              38.2    1.0    T2    N0    M0         distal  ...   \n",
       "GSM971961              39.2    4.0    T3    N1    M1         distal  ...   \n",
       "\n",
       "                      tp53.mutation.protein kras.mutation  kras.mutation.dna  \\\n",
       "!Sample_geo_accession                                                          \n",
       "GSM971957                               N/A             M            c.35G>A   \n",
       "GSM971958                               N/A            WT                N/A   \n",
       "GSM971959                               N/A            WT                N/A   \n",
       "GSM971960                               N/A            WT                N/A   \n",
       "GSM971961                               N/A            WT                N/A   \n",
       "\n",
       "                       kras.mutation.exon.number  kras.mutation.protein  \\\n",
       "!Sample_geo_accession                                                     \n",
       "GSM971957                                c.35G>A                 p.G12D   \n",
       "GSM971958                                    N/A                    N/A   \n",
       "GSM971959                                    N/A                    N/A   \n",
       "GSM971960                                    N/A                    N/A   \n",
       "GSM971961                                    N/A                    N/A   \n",
       "\n",
       "                       braf.mutation braf.mutation.dna  \\\n",
       "!Sample_geo_accession                                    \n",
       "GSM971957                         WT               NaN   \n",
       "GSM971958                         WT               NaN   \n",
       "GSM971959                         WT               NaN   \n",
       "GSM971960                         WT               NaN   \n",
       "GSM971961                         WT               NaN   \n",
       "\n",
       "                      braf.mutation.exon.number braf.mutation.protein  \\\n",
       "!Sample_geo_accession                                                   \n",
       "GSM971957                                   NaN                   NaN   \n",
       "GSM971958                                   NaN                   NaN   \n",
       "GSM971959                                   NaN                   NaN   \n",
       "GSM971960                                   NaN                   NaN   \n",
       "GSM971961                                   NaN                   NaN   \n",
       "\n",
       "                      cit.molecularsubtype  \n",
       "!Sample_geo_accession                       \n",
       "GSM971957                               C5  \n",
       "GSM971958                               C4  \n",
       "GSM971959                               C3  \n",
       "GSM971960                               C1  \n",
       "GSM971961                               C4  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_tissue= pd.read_csv(\n",
    "    \"data/GSE39582_clinicaldata.txt\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=\"!Sample_geo_accession\"\n",
    ")\n",
    "clin_tissue.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(585, 32)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clin_tissue.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>age</th>\n",
       "      <th>stage</th>\n",
       "      <th>DFS</th>\n",
       "      <th>DFSM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM971957</th>\n",
       "      <td>1</td>\n",
       "      <td>34.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971958</th>\n",
       "      <td>1</td>\n",
       "      <td>36.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971959</th>\n",
       "      <td>1</td>\n",
       "      <td>36.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971960</th>\n",
       "      <td>1</td>\n",
       "      <td>38.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971961</th>\n",
       "      <td>1</td>\n",
       "      <td>39.2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681367</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681368</th>\n",
       "      <td>0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681369</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681370</th>\n",
       "      <td>0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681371</th>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>585 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       target   age  stage  DFS  DFSM\n",
       "!Sample_geo_accession                                \n",
       "GSM971957                   1  34.5    4.0  1.0   0.0\n",
       "GSM971958                   1  36.4    4.0  1.0   0.0\n",
       "GSM971959                   1  36.9    2.0  1.0   4.0\n",
       "GSM971960                   1  38.2    1.0  0.0  74.0\n",
       "GSM971961                   1  39.2    4.0  1.0   1.0\n",
       "...                       ...   ...    ...  ...   ...\n",
       "GSM1681367                  0  67.0    3.0  0.0  76.0\n",
       "GSM1681368                  0  47.0    3.0  0.0  81.0\n",
       "GSM1681369                  0  67.0    3.0  0.0  65.0\n",
       "GSM1681370                  0  77.0    2.0  0.0  74.0\n",
       "GSM1681371                  0  69.0    1.0  0.0  68.0\n",
       "\n",
       "[585 rows x 5 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_tissue[\"target\"] = [1 if i == \" colorectal cancer\" else 0 for i in clin_tissue[\"status\"]]\n",
    "clin_tissue = clin_tissue[[\"target\", \"age\", \"stage\", \"DFS\", \"DFSM\"]]\n",
    "clin_tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>age</th>\n",
       "      <th>stage</th>\n",
       "      <th>DFS</th>\n",
       "      <th>DFSM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM971957</th>\n",
       "      <td>1</td>\n",
       "      <td>34.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971958</th>\n",
       "      <td>1</td>\n",
       "      <td>36.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971959</th>\n",
       "      <td>1</td>\n",
       "      <td>36.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971960</th>\n",
       "      <td>1</td>\n",
       "      <td>38.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971961</th>\n",
       "      <td>1</td>\n",
       "      <td>39.2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681367</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681368</th>\n",
       "      <td>0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681369</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681370</th>\n",
       "      <td>0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681371</th>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>573 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       target   age  stage  DFS  DFSM\n",
       "!Sample_geo_accession                                \n",
       "GSM971957                   1  34.5    4.0  1.0   0.0\n",
       "GSM971958                   1  36.4    4.0  1.0   0.0\n",
       "GSM971959                   1  36.9    2.0  1.0   4.0\n",
       "GSM971960                   1  38.2    1.0  0.0  74.0\n",
       "GSM971961                   1  39.2    4.0  1.0   1.0\n",
       "...                       ...   ...    ...  ...   ...\n",
       "GSM1681367                  0  67.0    3.0  0.0  76.0\n",
       "GSM1681368                  0  47.0    3.0  0.0  81.0\n",
       "GSM1681369                  0  67.0    3.0  0.0  65.0\n",
       "GSM1681370                  0  77.0    2.0  0.0  74.0\n",
       "GSM1681371                  0  69.0    1.0  0.0  68.0\n",
       "\n",
       "[573 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_tissue.dropna(inplace=True)\n",
    "clin_tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "target\n",
       "1    556\n",
       "0     17\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_tissue[\"target\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM971957</th>\n",
       "      <th>GSM971958</th>\n",
       "      <th>GSM971959</th>\n",
       "      <th>GSM971960</th>\n",
       "      <th>GSM971961</th>\n",
       "      <th>GSM971962</th>\n",
       "      <th>GSM971963</th>\n",
       "      <th>GSM971964</th>\n",
       "      <th>GSM971965</th>\n",
       "      <th>GSM971966</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM1681362</th>\n",
       "      <th>GSM1681363</th>\n",
       "      <th>GSM1681364</th>\n",
       "      <th>GSM1681365</th>\n",
       "      <th>GSM1681366</th>\n",
       "      <th>GSM1681367</th>\n",
       "      <th>GSM1681368</th>\n",
       "      <th>GSM1681369</th>\n",
       "      <th>GSM1681370</th>\n",
       "      <th>GSM1681371</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>9.561278</td>\n",
       "      <td>9.195198</td>\n",
       "      <td>9.934422</td>\n",
       "      <td>10.266136</td>\n",
       "      <td>9.977434</td>\n",
       "      <td>9.872145</td>\n",
       "      <td>9.045366</td>\n",
       "      <td>10.003417</td>\n",
       "      <td>9.990172</td>\n",
       "      <td>10.217966</td>\n",
       "      <td>...</td>\n",
       "      <td>9.847167</td>\n",
       "      <td>9.733898</td>\n",
       "      <td>9.904499</td>\n",
       "      <td>10.028011</td>\n",
       "      <td>10.031179</td>\n",
       "      <td>10.075013</td>\n",
       "      <td>9.506676</td>\n",
       "      <td>9.882653</td>\n",
       "      <td>9.898011</td>\n",
       "      <td>9.743342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>7.490955</td>\n",
       "      <td>7.549949</td>\n",
       "      <td>8.417405</td>\n",
       "      <td>8.165186</td>\n",
       "      <td>6.858895</td>\n",
       "      <td>7.456900</td>\n",
       "      <td>7.294602</td>\n",
       "      <td>8.046336</td>\n",
       "      <td>7.363857</td>\n",
       "      <td>7.798451</td>\n",
       "      <td>...</td>\n",
       "      <td>7.082756</td>\n",
       "      <td>7.216274</td>\n",
       "      <td>6.403616</td>\n",
       "      <td>6.705587</td>\n",
       "      <td>6.475627</td>\n",
       "      <td>6.277295</td>\n",
       "      <td>7.476909</td>\n",
       "      <td>6.674055</td>\n",
       "      <td>6.915436</td>\n",
       "      <td>7.043323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>5.483608</td>\n",
       "      <td>5.154432</td>\n",
       "      <td>4.982266</td>\n",
       "      <td>4.478060</td>\n",
       "      <td>6.206341</td>\n",
       "      <td>5.469860</td>\n",
       "      <td>6.144213</td>\n",
       "      <td>4.773469</td>\n",
       "      <td>5.155557</td>\n",
       "      <td>5.511558</td>\n",
       "      <td>...</td>\n",
       "      <td>4.832537</td>\n",
       "      <td>4.459356</td>\n",
       "      <td>4.842576</td>\n",
       "      <td>4.357467</td>\n",
       "      <td>4.627195</td>\n",
       "      <td>4.831889</td>\n",
       "      <td>4.708240</td>\n",
       "      <td>5.171778</td>\n",
       "      <td>4.745477</td>\n",
       "      <td>4.628314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>7.376411</td>\n",
       "      <td>7.375940</td>\n",
       "      <td>7.230739</td>\n",
       "      <td>7.548688</td>\n",
       "      <td>7.832766</td>\n",
       "      <td>7.241331</td>\n",
       "      <td>6.999633</td>\n",
       "      <td>6.949579</td>\n",
       "      <td>7.124734</td>\n",
       "      <td>7.162731</td>\n",
       "      <td>...</td>\n",
       "      <td>7.771227</td>\n",
       "      <td>7.934798</td>\n",
       "      <td>7.864958</td>\n",
       "      <td>8.044350</td>\n",
       "      <td>7.952567</td>\n",
       "      <td>7.880743</td>\n",
       "      <td>7.491522</td>\n",
       "      <td>7.668325</td>\n",
       "      <td>7.830421</td>\n",
       "      <td>7.732807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>2.845071</td>\n",
       "      <td>2.655784</td>\n",
       "      <td>2.837759</td>\n",
       "      <td>2.749647</td>\n",
       "      <td>2.885677</td>\n",
       "      <td>2.780455</td>\n",
       "      <td>2.936785</td>\n",
       "      <td>2.775945</td>\n",
       "      <td>2.712995</td>\n",
       "      <td>2.768064</td>\n",
       "      <td>...</td>\n",
       "      <td>2.910340</td>\n",
       "      <td>2.778019</td>\n",
       "      <td>2.730654</td>\n",
       "      <td>2.718647</td>\n",
       "      <td>2.720240</td>\n",
       "      <td>2.782004</td>\n",
       "      <td>2.879964</td>\n",
       "      <td>2.931177</td>\n",
       "      <td>2.876531</td>\n",
       "      <td>2.933604</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 585 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM971957  GSM971958  GSM971959  GSM971960  GSM971961  GSM971962  \\\n",
       "ID_REF                                                                        \n",
       "1007_s_at   9.561278   9.195198   9.934422  10.266136   9.977434   9.872145   \n",
       "1053_at     7.490955   7.549949   8.417405   8.165186   6.858895   7.456900   \n",
       "117_at      5.483608   5.154432   4.982266   4.478060   6.206341   5.469860   \n",
       "121_at      7.376411   7.375940   7.230739   7.548688   7.832766   7.241331   \n",
       "1255_g_at   2.845071   2.655784   2.837759   2.749647   2.885677   2.780455   \n",
       "\n",
       "           GSM971963  GSM971964  GSM971965  GSM971966  ...  GSM1681362  \\\n",
       "ID_REF                                                 ...               \n",
       "1007_s_at   9.045366  10.003417   9.990172  10.217966  ...    9.847167   \n",
       "1053_at     7.294602   8.046336   7.363857   7.798451  ...    7.082756   \n",
       "117_at      6.144213   4.773469   5.155557   5.511558  ...    4.832537   \n",
       "121_at      6.999633   6.949579   7.124734   7.162731  ...    7.771227   \n",
       "1255_g_at   2.936785   2.775945   2.712995   2.768064  ...    2.910340   \n",
       "\n",
       "           GSM1681363  GSM1681364  GSM1681365  GSM1681366  GSM1681367  \\\n",
       "ID_REF                                                                  \n",
       "1007_s_at    9.733898    9.904499   10.028011   10.031179   10.075013   \n",
       "1053_at      7.216274    6.403616    6.705587    6.475627    6.277295   \n",
       "117_at       4.459356    4.842576    4.357467    4.627195    4.831889   \n",
       "121_at       7.934798    7.864958    8.044350    7.952567    7.880743   \n",
       "1255_g_at    2.778019    2.730654    2.718647    2.720240    2.782004   \n",
       "\n",
       "           GSM1681368  GSM1681369  GSM1681370  GSM1681371  \n",
       "ID_REF                                                     \n",
       "1007_s_at    9.506676    9.882653    9.898011    9.743342  \n",
       "1053_at      7.476909    6.674055    6.915436    7.043323  \n",
       "117_at       4.708240    5.171778    4.745477    4.628314  \n",
       "121_at       7.491522    7.668325    7.830421    7.732807  \n",
       "1255_g_at    2.879964    2.931177    2.876531    2.933604  \n",
       "\n",
       "[5 rows x 585 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_tissue= pd.read_csv(\n",
    "    \"data/GSE39582_genes.txt\", \n",
    "    sep=\"\\t\",\n",
    "    index_col=\"ID_REF\"\n",
    ")\n",
    "genes_tissue.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>age</th>\n",
       "      <th>stage</th>\n",
       "      <th>DFS</th>\n",
       "      <th>DFSM</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM971957</th>\n",
       "      <td>1</td>\n",
       "      <td>34.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.561278</td>\n",
       "      <td>7.490955</td>\n",
       "      <td>5.483608</td>\n",
       "      <td>7.376411</td>\n",
       "      <td>2.845071</td>\n",
       "      <td>...</td>\n",
       "      <td>11.475950</td>\n",
       "      <td>10.483983</td>\n",
       "      <td>12.623022</td>\n",
       "      <td>12.593145</td>\n",
       "      <td>3.547751</td>\n",
       "      <td>3.929784</td>\n",
       "      <td>3.058311</td>\n",
       "      <td>2.697554</td>\n",
       "      <td>3.038737</td>\n",
       "      <td>2.802092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971958</th>\n",
       "      <td>1</td>\n",
       "      <td>36.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.195198</td>\n",
       "      <td>7.549949</td>\n",
       "      <td>5.154432</td>\n",
       "      <td>7.375940</td>\n",
       "      <td>2.655784</td>\n",
       "      <td>...</td>\n",
       "      <td>11.572610</td>\n",
       "      <td>10.807687</td>\n",
       "      <td>13.093898</td>\n",
       "      <td>13.047588</td>\n",
       "      <td>3.545181</td>\n",
       "      <td>3.309591</td>\n",
       "      <td>2.691952</td>\n",
       "      <td>2.434740</td>\n",
       "      <td>2.824796</td>\n",
       "      <td>2.600676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971959</th>\n",
       "      <td>1</td>\n",
       "      <td>36.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.934422</td>\n",
       "      <td>8.417405</td>\n",
       "      <td>4.982266</td>\n",
       "      <td>7.230739</td>\n",
       "      <td>2.837759</td>\n",
       "      <td>...</td>\n",
       "      <td>11.656232</td>\n",
       "      <td>10.864199</td>\n",
       "      <td>12.989416</td>\n",
       "      <td>13.188809</td>\n",
       "      <td>4.014747</td>\n",
       "      <td>4.114073</td>\n",
       "      <td>3.189148</td>\n",
       "      <td>2.713932</td>\n",
       "      <td>3.182882</td>\n",
       "      <td>2.749798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971960</th>\n",
       "      <td>1</td>\n",
       "      <td>38.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>10.266136</td>\n",
       "      <td>8.165186</td>\n",
       "      <td>4.478060</td>\n",
       "      <td>7.548688</td>\n",
       "      <td>2.749647</td>\n",
       "      <td>...</td>\n",
       "      <td>11.139844</td>\n",
       "      <td>10.278899</td>\n",
       "      <td>12.684215</td>\n",
       "      <td>12.569020</td>\n",
       "      <td>3.994478</td>\n",
       "      <td>3.259446</td>\n",
       "      <td>3.361405</td>\n",
       "      <td>2.551093</td>\n",
       "      <td>3.247612</td>\n",
       "      <td>2.926161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971961</th>\n",
       "      <td>1</td>\n",
       "      <td>39.2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.977434</td>\n",
       "      <td>6.858895</td>\n",
       "      <td>6.206341</td>\n",
       "      <td>7.832766</td>\n",
       "      <td>2.885677</td>\n",
       "      <td>...</td>\n",
       "      <td>11.845596</td>\n",
       "      <td>10.868646</td>\n",
       "      <td>13.020133</td>\n",
       "      <td>13.042203</td>\n",
       "      <td>3.540059</td>\n",
       "      <td>3.675026</td>\n",
       "      <td>2.683685</td>\n",
       "      <td>2.624718</td>\n",
       "      <td>3.120898</td>\n",
       "      <td>2.560125</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 54680 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           target   age  stage  DFS  DFSM  1007_s_at   1053_at    117_at  \\\n",
       "GSM971957       1  34.5    4.0  1.0   0.0   9.561278  7.490955  5.483608   \n",
       "GSM971958       1  36.4    4.0  1.0   0.0   9.195198  7.549949  5.154432   \n",
       "GSM971959       1  36.9    2.0  1.0   4.0   9.934422  8.417405  4.982266   \n",
       "GSM971960       1  38.2    1.0  0.0  74.0  10.266136  8.165186  4.478060   \n",
       "GSM971961       1  39.2    4.0  1.0   1.0   9.977434  6.858895  6.206341   \n",
       "\n",
       "             121_at  1255_g_at  ...  AFFX-r2-Ec-bioD-3_at  \\\n",
       "GSM971957  7.376411   2.845071  ...             11.475950   \n",
       "GSM971958  7.375940   2.655784  ...             11.572610   \n",
       "GSM971959  7.230739   2.837759  ...             11.656232   \n",
       "GSM971960  7.548688   2.749647  ...             11.139844   \n",
       "GSM971961  7.832766   2.885677  ...             11.845596   \n",
       "\n",
       "           AFFX-r2-Ec-bioD-5_at  AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "GSM971957             10.483983            12.623022            12.593145   \n",
       "GSM971958             10.807687            13.093898            13.047588   \n",
       "GSM971959             10.864199            12.989416            13.188809   \n",
       "GSM971960             10.278899            12.684215            12.569020   \n",
       "GSM971961             10.868646            13.020133            13.042203   \n",
       "\n",
       "           AFFX-ThrX-3_at  AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  \\\n",
       "GSM971957        3.547751        3.929784        3.058311         2.697554   \n",
       "GSM971958        3.545181        3.309591        2.691952         2.434740   \n",
       "GSM971959        4.014747        4.114073        3.189148         2.713932   \n",
       "GSM971960        3.994478        3.259446        3.361405         2.551093   \n",
       "GSM971961        3.540059        3.675026        2.683685         2.624718   \n",
       "\n",
       "           AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "GSM971957         3.038737         2.802092  \n",
       "GSM971958         2.824796         2.600676  \n",
       "GSM971959         3.182882         2.749798  \n",
       "GSM971960         3.247612         2.926161  \n",
       "GSM971961         3.120898         2.560125  \n",
       "\n",
       "[5 rows x 54680 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.merge(clin_tissue, genes_tissue.T, left_index=True, right_index=True)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(573, 54680)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blood Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_title</th>\n",
       "      <th>!Sample_source_name_ch1</th>\n",
       "      <th>!Sample_organism_ch1</th>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <th>!Sample_characteristics_ch1.1</th>\n",
       "      <th>status</th>\n",
       "      <th>!Sample_characteristics_ch1.2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5001936</th>\n",
       "      <td>006-ctr016: Peripheral Blood_Healthy control</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>age: 43</td>\n",
       "      <td>normal</td>\n",
       "      <td>tissue: Peripheral Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001937</th>\n",
       "      <td>006-ctr021: Peripheral Blood_Healthy control</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>age: 47</td>\n",
       "      <td>normal</td>\n",
       "      <td>tissue: Peripheral Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001938</th>\n",
       "      <td>006-ctr030: Peripheral Blood_Healthy control</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>age: 42</td>\n",
       "      <td>normal</td>\n",
       "      <td>tissue: Peripheral Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001939</th>\n",
       "      <td>006-ctr057: Peripheral Blood_Healthy control</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>gender: Female</td>\n",
       "      <td>age: 44</td>\n",
       "      <td>normal</td>\n",
       "      <td>tissue: Peripheral Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001940</th>\n",
       "      <td>006-ctr065: Peripheral Blood_Healthy control</td>\n",
       "      <td>Peripheral Blood</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>gender: Male</td>\n",
       "      <td>age: 55</td>\n",
       "      <td>normal</td>\n",
       "      <td>tissue: Peripheral Blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      !Sample_title  \\\n",
       "!Sample_geo_accession                                                 \n",
       "GSM5001936             006-ctr016: Peripheral Blood_Healthy control   \n",
       "GSM5001937             006-ctr021: Peripheral Blood_Healthy control   \n",
       "GSM5001938             006-ctr030: Peripheral Blood_Healthy control   \n",
       "GSM5001939             006-ctr057: Peripheral Blood_Healthy control   \n",
       "GSM5001940             006-ctr065: Peripheral Blood_Healthy control   \n",
       "\n",
       "                      !Sample_source_name_ch1 !Sample_organism_ch1  \\\n",
       "!Sample_geo_accession                                                \n",
       "GSM5001936                   Peripheral Blood         Homo sapiens   \n",
       "GSM5001937                   Peripheral Blood         Homo sapiens   \n",
       "GSM5001938                   Peripheral Blood         Homo sapiens   \n",
       "GSM5001939                   Peripheral Blood         Homo sapiens   \n",
       "GSM5001940                   Peripheral Blood         Homo sapiens   \n",
       "\n",
       "                      !Sample_characteristics_ch1  \\\n",
       "!Sample_geo_accession                               \n",
       "GSM5001936                           gender: Male   \n",
       "GSM5001937                           gender: Male   \n",
       "GSM5001938                         gender: Female   \n",
       "GSM5001939                         gender: Female   \n",
       "GSM5001940                           gender: Male   \n",
       "\n",
       "                      !Sample_characteristics_ch1.1   status  \\\n",
       "!Sample_geo_accession                                          \n",
       "GSM5001936                                  age: 43   normal   \n",
       "GSM5001937                                  age: 47   normal   \n",
       "GSM5001938                                  age: 42   normal   \n",
       "GSM5001939                                  age: 44   normal   \n",
       "GSM5001940                                  age: 55   normal   \n",
       "\n",
       "                      !Sample_characteristics_ch1.2  \n",
       "!Sample_geo_accession                                \n",
       "GSM5001936                 tissue: Peripheral Blood  \n",
       "GSM5001937                 tissue: Peripheral Blood  \n",
       "GSM5001938                 tissue: Peripheral Blood  \n",
       "GSM5001939                 tissue: Peripheral Blood  \n",
       "GSM5001940                 tissue: Peripheral Blood  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_blood= pd.read_csv(\n",
    "    \"data/GSE164191_clinicaldata.txt\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=\"!Sample_geo_accession\"\n",
    ")\n",
    "clin_blood.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(121, 7)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_blood.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{' colorectal cancer', ' normal'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(clin_blood[\"status\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5001936</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001937</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001938</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001939</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001940</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5002052</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5002053</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5002054</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5002055</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5002056</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       target\n",
       "!Sample_geo_accession        \n",
       "GSM5001936                  0\n",
       "GSM5001937                  0\n",
       "GSM5001938                  0\n",
       "GSM5001939                  0\n",
       "GSM5001940                  0\n",
       "...                       ...\n",
       "GSM5002052                  1\n",
       "GSM5002053                  1\n",
       "GSM5002054                  1\n",
       "GSM5002055                  1\n",
       "GSM5002056                  1\n",
       "\n",
       "[121 rows x 1 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_blood[\"target\"] = [1 if i == \" colorectal cancer\" else 0 for i in clin_blood[\"status\"]]\n",
    "clin_blood = clin_blood[[\"target\"]]\n",
    "clin_blood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "target\n",
       "0    62\n",
       "1    59\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_blood.value_counts(\"target\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM5001936</th>\n",
       "      <th>GSM5001937</th>\n",
       "      <th>GSM5001938</th>\n",
       "      <th>GSM5001939</th>\n",
       "      <th>GSM5001940</th>\n",
       "      <th>GSM5001941</th>\n",
       "      <th>GSM5001942</th>\n",
       "      <th>GSM5001943</th>\n",
       "      <th>GSM5001944</th>\n",
       "      <th>GSM5001945</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM5002047</th>\n",
       "      <th>GSM5002048</th>\n",
       "      <th>GSM5002049</th>\n",
       "      <th>GSM5002050</th>\n",
       "      <th>GSM5002051</th>\n",
       "      <th>GSM5002052</th>\n",
       "      <th>GSM5002053</th>\n",
       "      <th>GSM5002054</th>\n",
       "      <th>GSM5002055</th>\n",
       "      <th>GSM5002056</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007_s_at</th>\n",
       "      <td>725.8980</td>\n",
       "      <td>796.0900</td>\n",
       "      <td>512.20100</td>\n",
       "      <td>1073.3500</td>\n",
       "      <td>911.0790</td>\n",
       "      <td>723.7810</td>\n",
       "      <td>853.90600</td>\n",
       "      <td>797.07600</td>\n",
       "      <td>447.0500</td>\n",
       "      <td>628.674</td>\n",
       "      <td>...</td>\n",
       "      <td>989.9150</td>\n",
       "      <td>771.9320</td>\n",
       "      <td>560.6720</td>\n",
       "      <td>909.0870</td>\n",
       "      <td>537.1380</td>\n",
       "      <td>950.6690</td>\n",
       "      <td>1401.660</td>\n",
       "      <td>1086.33000</td>\n",
       "      <td>1073.8700</td>\n",
       "      <td>1053.52000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1053_at</th>\n",
       "      <td>395.5860</td>\n",
       "      <td>399.6040</td>\n",
       "      <td>387.95500</td>\n",
       "      <td>370.9710</td>\n",
       "      <td>470.6210</td>\n",
       "      <td>382.3690</td>\n",
       "      <td>307.47900</td>\n",
       "      <td>355.33600</td>\n",
       "      <td>426.2140</td>\n",
       "      <td>400.152</td>\n",
       "      <td>...</td>\n",
       "      <td>468.1380</td>\n",
       "      <td>339.8340</td>\n",
       "      <td>444.1970</td>\n",
       "      <td>277.7600</td>\n",
       "      <td>433.6500</td>\n",
       "      <td>441.2880</td>\n",
       "      <td>482.165</td>\n",
       "      <td>281.64000</td>\n",
       "      <td>362.3690</td>\n",
       "      <td>459.42000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117_at</th>\n",
       "      <td>2733.8700</td>\n",
       "      <td>2608.8000</td>\n",
       "      <td>2322.60000</td>\n",
       "      <td>2639.6300</td>\n",
       "      <td>3788.7800</td>\n",
       "      <td>4272.0200</td>\n",
       "      <td>2723.17000</td>\n",
       "      <td>2084.72000</td>\n",
       "      <td>4792.2900</td>\n",
       "      <td>2605.910</td>\n",
       "      <td>...</td>\n",
       "      <td>2862.3600</td>\n",
       "      <td>2175.9300</td>\n",
       "      <td>2152.0700</td>\n",
       "      <td>1599.2100</td>\n",
       "      <td>2378.1300</td>\n",
       "      <td>1448.0500</td>\n",
       "      <td>2308.200</td>\n",
       "      <td>1598.52000</td>\n",
       "      <td>3011.1000</td>\n",
       "      <td>2247.66000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121_at</th>\n",
       "      <td>1440.2100</td>\n",
       "      <td>1500.6300</td>\n",
       "      <td>1262.00000</td>\n",
       "      <td>1547.6700</td>\n",
       "      <td>1092.4300</td>\n",
       "      <td>1104.6600</td>\n",
       "      <td>1393.71000</td>\n",
       "      <td>948.32200</td>\n",
       "      <td>676.1600</td>\n",
       "      <td>1332.270</td>\n",
       "      <td>...</td>\n",
       "      <td>1695.9300</td>\n",
       "      <td>829.6490</td>\n",
       "      <td>1411.2200</td>\n",
       "      <td>1510.9200</td>\n",
       "      <td>1765.7500</td>\n",
       "      <td>1642.1500</td>\n",
       "      <td>1986.370</td>\n",
       "      <td>1558.57000</td>\n",
       "      <td>1989.4400</td>\n",
       "      <td>2080.07000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255_g_at</th>\n",
       "      <td>38.3712</td>\n",
       "      <td>39.1397</td>\n",
       "      <td>5.62705</td>\n",
       "      <td>10.2502</td>\n",
       "      <td>58.7812</td>\n",
       "      <td>35.7721</td>\n",
       "      <td>6.09315</td>\n",
       "      <td>4.19176</td>\n",
       "      <td>22.2372</td>\n",
       "      <td>79.154</td>\n",
       "      <td>...</td>\n",
       "      <td>82.0326</td>\n",
       "      <td>32.7375</td>\n",
       "      <td>23.0333</td>\n",
       "      <td>39.0699</td>\n",
       "      <td>18.0678</td>\n",
       "      <td>12.3742</td>\n",
       "      <td>118.104</td>\n",
       "      <td>9.95137</td>\n",
       "      <td>56.7037</td>\n",
       "      <td>9.38501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 121 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GSM5001936  GSM5001937  GSM5001938  GSM5001939  GSM5001940  \\\n",
       "ID_REF                                                                  \n",
       "1007_s_at    725.8980    796.0900   512.20100   1073.3500    911.0790   \n",
       "1053_at      395.5860    399.6040   387.95500    370.9710    470.6210   \n",
       "117_at      2733.8700   2608.8000  2322.60000   2639.6300   3788.7800   \n",
       "121_at      1440.2100   1500.6300  1262.00000   1547.6700   1092.4300   \n",
       "1255_g_at     38.3712     39.1397     5.62705     10.2502     58.7812   \n",
       "\n",
       "           GSM5001941  GSM5001942  GSM5001943  GSM5001944  GSM5001945  ...  \\\n",
       "ID_REF                                                                 ...   \n",
       "1007_s_at    723.7810   853.90600   797.07600    447.0500     628.674  ...   \n",
       "1053_at      382.3690   307.47900   355.33600    426.2140     400.152  ...   \n",
       "117_at      4272.0200  2723.17000  2084.72000   4792.2900    2605.910  ...   \n",
       "121_at      1104.6600  1393.71000   948.32200    676.1600    1332.270  ...   \n",
       "1255_g_at     35.7721     6.09315     4.19176     22.2372      79.154  ...   \n",
       "\n",
       "           GSM5002047  GSM5002048  GSM5002049  GSM5002050  GSM5002051  \\\n",
       "ID_REF                                                                  \n",
       "1007_s_at    989.9150    771.9320    560.6720    909.0870    537.1380   \n",
       "1053_at      468.1380    339.8340    444.1970    277.7600    433.6500   \n",
       "117_at      2862.3600   2175.9300   2152.0700   1599.2100   2378.1300   \n",
       "121_at      1695.9300    829.6490   1411.2200   1510.9200   1765.7500   \n",
       "1255_g_at     82.0326     32.7375     23.0333     39.0699     18.0678   \n",
       "\n",
       "           GSM5002052  GSM5002053  GSM5002054  GSM5002055  GSM5002056  \n",
       "ID_REF                                                                 \n",
       "1007_s_at    950.6690    1401.660  1086.33000   1073.8700  1053.52000  \n",
       "1053_at      441.2880     482.165   281.64000    362.3690   459.42000  \n",
       "117_at      1448.0500    2308.200  1598.52000   3011.1000  2247.66000  \n",
       "121_at      1642.1500    1986.370  1558.57000   1989.4400  2080.07000  \n",
       "1255_g_at     12.3742     118.104     9.95137     56.7037     9.38501  \n",
       "\n",
       "[5 rows x 121 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_blood= pd.read_csv(\n",
    "    \"data/GSE164191_genes.txt\", \n",
    "    sep=\"\\t\",\n",
    "    index_col=\"ID_REF\"\n",
    ")\n",
    "genes_blood.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54675, 121)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_blood.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54675"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from process_func import filter_iqr\n",
    "keep_gene_lst = filter_iqr(genes_blood, 0.5)\n",
    "len(keep_gene_lst)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### same number with previous --> blood data doesnt have lowly expressed genes, so we should filter by tissue (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23711"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_gene_lst = filter_iqr(genes_tissue, 0.5)\n",
    "len(keep_gene_lst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23711, 121)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_blood = genes_blood.loc[keep_gene_lst]\n",
    "genes_blood.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "      <th>121_at</th>\n",
       "      <th>1255_g_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1316_at</th>\n",
       "      <th>1320_at</th>\n",
       "      <th>1405_i_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-Ec-bioD-3_at</th>\n",
       "      <th>AFFX-r2-Ec-bioD-5_at</th>\n",
       "      <th>AFFX-r2-P1-cre-3_at</th>\n",
       "      <th>AFFX-r2-P1-cre-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5001936</th>\n",
       "      <td>0</td>\n",
       "      <td>725.898</td>\n",
       "      <td>395.586</td>\n",
       "      <td>2733.87</td>\n",
       "      <td>1440.21</td>\n",
       "      <td>38.37120</td>\n",
       "      <td>1259.94</td>\n",
       "      <td>282.894</td>\n",
       "      <td>22.32070</td>\n",
       "      <td>23357.6</td>\n",
       "      <td>...</td>\n",
       "      <td>32278.4</td>\n",
       "      <td>19026.7</td>\n",
       "      <td>83572.0</td>\n",
       "      <td>59173.0</td>\n",
       "      <td>927.437</td>\n",
       "      <td>406.163</td>\n",
       "      <td>580.921</td>\n",
       "      <td>42.1028</td>\n",
       "      <td>81.7244</td>\n",
       "      <td>9.91339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001937</th>\n",
       "      <td>0</td>\n",
       "      <td>796.090</td>\n",
       "      <td>399.604</td>\n",
       "      <td>2608.80</td>\n",
       "      <td>1500.63</td>\n",
       "      <td>39.13970</td>\n",
       "      <td>1121.50</td>\n",
       "      <td>206.749</td>\n",
       "      <td>95.31110</td>\n",
       "      <td>20629.3</td>\n",
       "      <td>...</td>\n",
       "      <td>35960.7</td>\n",
       "      <td>22217.7</td>\n",
       "      <td>94596.8</td>\n",
       "      <td>65353.5</td>\n",
       "      <td>1360.550</td>\n",
       "      <td>646.118</td>\n",
       "      <td>969.375</td>\n",
       "      <td>51.2552</td>\n",
       "      <td>16.7812</td>\n",
       "      <td>5.35075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001938</th>\n",
       "      <td>0</td>\n",
       "      <td>512.201</td>\n",
       "      <td>387.955</td>\n",
       "      <td>2322.60</td>\n",
       "      <td>1262.00</td>\n",
       "      <td>5.62705</td>\n",
       "      <td>1126.70</td>\n",
       "      <td>347.664</td>\n",
       "      <td>12.37100</td>\n",
       "      <td>27487.3</td>\n",
       "      <td>...</td>\n",
       "      <td>27478.8</td>\n",
       "      <td>17796.6</td>\n",
       "      <td>75012.9</td>\n",
       "      <td>52527.9</td>\n",
       "      <td>1149.930</td>\n",
       "      <td>623.787</td>\n",
       "      <td>938.115</td>\n",
       "      <td>36.6352</td>\n",
       "      <td>38.5536</td>\n",
       "      <td>3.54501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001939</th>\n",
       "      <td>0</td>\n",
       "      <td>1073.350</td>\n",
       "      <td>370.971</td>\n",
       "      <td>2639.63</td>\n",
       "      <td>1547.67</td>\n",
       "      <td>10.25020</td>\n",
       "      <td>1708.83</td>\n",
       "      <td>380.185</td>\n",
       "      <td>6.04854</td>\n",
       "      <td>40321.1</td>\n",
       "      <td>...</td>\n",
       "      <td>62129.5</td>\n",
       "      <td>38590.3</td>\n",
       "      <td>187277.0</td>\n",
       "      <td>111348.0</td>\n",
       "      <td>383.647</td>\n",
       "      <td>117.674</td>\n",
       "      <td>223.904</td>\n",
       "      <td>24.8811</td>\n",
       "      <td>97.5604</td>\n",
       "      <td>11.61090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5001940</th>\n",
       "      <td>0</td>\n",
       "      <td>911.079</td>\n",
       "      <td>470.621</td>\n",
       "      <td>3788.78</td>\n",
       "      <td>1092.43</td>\n",
       "      <td>58.78120</td>\n",
       "      <td>1173.27</td>\n",
       "      <td>206.402</td>\n",
       "      <td>34.44670</td>\n",
       "      <td>14319.7</td>\n",
       "      <td>...</td>\n",
       "      <td>31335.8</td>\n",
       "      <td>19608.5</td>\n",
       "      <td>89518.4</td>\n",
       "      <td>60446.0</td>\n",
       "      <td>791.052</td>\n",
       "      <td>211.981</td>\n",
       "      <td>403.888</td>\n",
       "      <td>10.6119</td>\n",
       "      <td>30.3497</td>\n",
       "      <td>4.82076</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 54676 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       target  1007_s_at  1053_at   117_at   121_at  \\\n",
       "!Sample_geo_accession                                                 \n",
       "GSM5001936                  0    725.898  395.586  2733.87  1440.21   \n",
       "GSM5001937                  0    796.090  399.604  2608.80  1500.63   \n",
       "GSM5001938                  0    512.201  387.955  2322.60  1262.00   \n",
       "GSM5001939                  0   1073.350  370.971  2639.63  1547.67   \n",
       "GSM5001940                  0    911.079  470.621  3788.78  1092.43   \n",
       "\n",
       "                       1255_g_at  1294_at  1316_at   1320_at  1405_i_at  ...  \\\n",
       "!Sample_geo_accession                                                    ...   \n",
       "GSM5001936              38.37120  1259.94  282.894  22.32070    23357.6  ...   \n",
       "GSM5001937              39.13970  1121.50  206.749  95.31110    20629.3  ...   \n",
       "GSM5001938               5.62705  1126.70  347.664  12.37100    27487.3  ...   \n",
       "GSM5001939              10.25020  1708.83  380.185   6.04854    40321.1  ...   \n",
       "GSM5001940              58.78120  1173.27  206.402  34.44670    14319.7  ...   \n",
       "\n",
       "                       AFFX-r2-Ec-bioD-3_at  AFFX-r2-Ec-bioD-5_at  \\\n",
       "!Sample_geo_accession                                               \n",
       "GSM5001936                          32278.4               19026.7   \n",
       "GSM5001937                          35960.7               22217.7   \n",
       "GSM5001938                          27478.8               17796.6   \n",
       "GSM5001939                          62129.5               38590.3   \n",
       "GSM5001940                          31335.8               19608.5   \n",
       "\n",
       "                       AFFX-r2-P1-cre-3_at  AFFX-r2-P1-cre-5_at  \\\n",
       "!Sample_geo_accession                                             \n",
       "GSM5001936                         83572.0              59173.0   \n",
       "GSM5001937                         94596.8              65353.5   \n",
       "GSM5001938                         75012.9              52527.9   \n",
       "GSM5001939                        187277.0             111348.0   \n",
       "GSM5001940                         89518.4              60446.0   \n",
       "\n",
       "                       AFFX-ThrX-3_at  AFFX-ThrX-5_at  AFFX-ThrX-M_at  \\\n",
       "!Sample_geo_accession                                                   \n",
       "GSM5001936                    927.437         406.163         580.921   \n",
       "GSM5001937                   1360.550         646.118         969.375   \n",
       "GSM5001938                   1149.930         623.787         938.115   \n",
       "GSM5001939                    383.647         117.674         223.904   \n",
       "GSM5001940                    791.052         211.981         403.888   \n",
       "\n",
       "                       AFFX-TrpnX-3_at  AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "!Sample_geo_accession                                                     \n",
       "GSM5001936                     42.1028          81.7244          9.91339  \n",
       "GSM5001937                     51.2552          16.7812          5.35075  \n",
       "GSM5001938                     36.6352          38.5536          3.54501  \n",
       "GSM5001939                     24.8811          97.5604         11.61090  \n",
       "GSM5001940                     10.6119          30.3497          4.82076  \n",
       "\n",
       "[5 rows x 54676 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.merge(clin_blood, genes_blood.T, left_index=True, right_index=True)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from process_func import select_features, model_name\n",
    "\n",
    "# Select features based on ANOVA f-test\n",
    "features = select_features(\n",
    "    data,\n",
    "    200,\n",
    "    0.5,\n",
    "    \"Standard\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    \"224963_at\" in features\n",
    "    and \"227331_at\" in features\n",
    "    and \"230085_at\" in features\n",
    "    and \"212514_x_at\" in features\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MAKING COMBINATION...\n",
      "\n",
      "\n",
      "ADD FEATURES\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1333500/1333500 [00:00<00:00, 1706096.93it/s]\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "\n",
    "# Combine selected features to create model names -> 200 features -> more than 1M models will have to be built\n",
    "print(\"\\nMAKING COMBINATION...\\n\")\n",
    "combinations_all = []\n",
    "for i in range(1, 3+1):\n",
    "    combinations_all.extend(list(itertools.combinations(features, i)))\n",
    "features = combinations_all\n",
    "print(\"\\nADD FEATURES\\n\")\n",
    "featurelist = []\n",
    "for i in tqdm(features):\n",
    "    selected_features = []\n",
    "    for k in i:\n",
    "        selected_features.append(k)\n",
    "    feature_name = model_name(selected_features)\n",
    "    featurelist.append(feature_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"featurelist.csv\", \"w\") as file:\n",
    "    file.write(\",\".join(featurelist))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting imbalanced-learn\n",
      "  Downloading imbalanced_learn-0.12.4-py3-none-any.whl.metadata (8.3 kB)\n",
      "Requirement already satisfied: numpy>=1.17.3 in c:\\users\\minh\\miniconda3\\envs\\crc_biomarker\\lib\\site-packages (from imbalanced-learn) (1.26.4)\n",
      "Requirement already satisfied: scipy>=1.5.0 in c:\\users\\minh\\miniconda3\\envs\\crc_biomarker\\lib\\site-packages (from imbalanced-learn) (1.14.1)\n",
      "Requirement already satisfied: scikit-learn>=1.0.2 in c:\\users\\minh\\miniconda3\\envs\\crc_biomarker\\lib\\site-packages (from imbalanced-learn) (1.5.2)\n",
      "Requirement already satisfied: joblib>=1.1.1 in c:\\users\\minh\\miniconda3\\envs\\crc_biomarker\\lib\\site-packages (from imbalanced-learn) (1.4.2)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\minh\\miniconda3\\envs\\crc_biomarker\\lib\\site-packages (from imbalanced-learn) (3.5.0)\n",
      "Downloading imbalanced_learn-0.12.4-py3-none-any.whl (258 kB)\n",
      "Installing collected packages: imbalanced-learn\n",
      "Successfully installed imbalanced-learn-0.12.4\n"
     ]
    }
   ],
   "source": [
    "!pip install imbalanced-learn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>age</th>\n",
       "      <th>stage</th>\n",
       "      <th>DFS</th>\n",
       "      <th>DFSM</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1053_at</th>\n",
       "      <th>117_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM971957</th>\n",
       "      <td>1</td>\n",
       "      <td>34.5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.561278</td>\n",
       "      <td>7.490955</td>\n",
       "      <td>5.483608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971958</th>\n",
       "      <td>1</td>\n",
       "      <td>36.4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.195198</td>\n",
       "      <td>7.549949</td>\n",
       "      <td>5.154432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971959</th>\n",
       "      <td>1</td>\n",
       "      <td>36.9</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.934422</td>\n",
       "      <td>8.417405</td>\n",
       "      <td>4.982266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971960</th>\n",
       "      <td>1</td>\n",
       "      <td>38.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>10.266136</td>\n",
       "      <td>8.165186</td>\n",
       "      <td>4.478060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM971961</th>\n",
       "      <td>1</td>\n",
       "      <td>39.2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.977434</td>\n",
       "      <td>6.858895</td>\n",
       "      <td>6.206341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681367</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>10.075013</td>\n",
       "      <td>6.277295</td>\n",
       "      <td>4.831889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681368</th>\n",
       "      <td>0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>9.506676</td>\n",
       "      <td>7.476909</td>\n",
       "      <td>4.708240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681369</th>\n",
       "      <td>0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>9.882653</td>\n",
       "      <td>6.674055</td>\n",
       "      <td>5.171778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681370</th>\n",
       "      <td>0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>9.898011</td>\n",
       "      <td>6.915436</td>\n",
       "      <td>4.745477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1681371</th>\n",
       "      <td>0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>9.743342</td>\n",
       "      <td>7.043323</td>\n",
       "      <td>4.628314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>573 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            target   age  stage  DFS  DFSM  1007_s_at   1053_at    117_at\n",
       "GSM971957        1  34.5    4.0  1.0   0.0   9.561278  7.490955  5.483608\n",
       "GSM971958        1  36.4    4.0  1.0   0.0   9.195198  7.549949  5.154432\n",
       "GSM971959        1  36.9    2.0  1.0   4.0   9.934422  8.417405  4.982266\n",
       "GSM971960        1  38.2    1.0  0.0  74.0  10.266136  8.165186  4.478060\n",
       "GSM971961        1  39.2    4.0  1.0   1.0   9.977434  6.858895  6.206341\n",
       "...            ...   ...    ...  ...   ...        ...       ...       ...\n",
       "GSM1681367       0  67.0    3.0  0.0  76.0  10.075013  6.277295  4.831889\n",
       "GSM1681368       0  47.0    3.0  0.0  81.0   9.506676  7.476909  4.708240\n",
       "GSM1681369       0  67.0    3.0  0.0  65.0   9.882653  6.674055  5.171778\n",
       "GSM1681370       0  77.0    2.0  0.0  74.0   9.898011  6.915436  4.745477\n",
       "GSM1681371       0  69.0    1.0  0.0  68.0   9.743342  7.043323  4.628314\n",
       "\n",
       "[573 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\n",
    "    [\"target\", \"age\", \"stage\", \"DFS\", \"DFSM\"] +\n",
    "    [\"1007_s_at\", \"1053_at\", \"117_at\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from imblearn.over_sampling import ADASYN\n",
    "adasyn = ADASYN(random_state=42)\n",
    "X, y = adasyn.fit_resample(\n",
    "    data[[\"DFS\", \"DFSM\", \"1007_s_at\", \"1053_at\", \"117_at\"]],\n",
    "    data[[\"target\"]]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"used_featurelist.csv\", \"r\") as file:\n",
    "    content = file.read()\n",
    "    features = content.split(\",\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "assm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
